psh/date/date.c

```diff
@@ -0,0 +1,153 @@
+/*
+ * Phoenix-RTOS
+ *
+ * date - print or set the system date and time
+ *
+ * Copyright 2022 Phoenix Systems
+ * Author: Mateusz Niewiadomski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+#include <sys/time.h>
+#include <unistd.h>
+
+#include "../psh.h"
+
+
+char *wdaylookup[] = { "Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat" };
+
+char *monlookup[] = { "Jan", "Feb", "Mar", "Apr", "May", "June", "July", "Aug", "Sept", "Oct", "Nov", "Dec" };
+
+enum mode { SKIP,
+	HELP,
+	PRINT,
+	WRITE };
+
+static void psh_date_info(void)
+{
+	printf("print/set the system date and time");
+}
+
+
+static void psh_date_help(const char *prog)
+{
+	printf("Usage: %s [option] [epoch]\n", prog);
+	printf("  -h:  shows this help message\n");
+	printf("  -s:  set system time described by EPOCH (POSIX time format)\n");
+}
+
+
+static int psh_date_print(void)
+{
+	time_t t;
+	struct tm *tinfo;
+
+	time(&t);
+	tinfo = localtime(&t);
+
+	printf("%s,", wdaylookup[tinfo->tm_wday]);
+	/* `struct tm` field tm_year is years since 1900, so modulo for last 2 digits */
+	printf(" %02d %s %d", tinfo->tm_mday, monlookup[tinfo->tm_mon], (tinfo->tm_year) % 100);
+	printf(" %02d:%02d:%02d", tinfo->tm_hour, tinfo->tm_min, tinfo->tm_sec);
+	printf("\n");
+
+	return 0;
+}
+
+
+static int psh_date_set(char *timestring)
+{
+	struct timeval tv = { .tv_sec = 0, .tv_usec = 0 };
+	struct timezone tz = { .tz_dsttime = 0, .tz_minuteswest = 0 };
+
+	if (timestring == NULL) {
+		return 1;
+	}
+
+	if (strlen(timestring) == 1 && timestring[0] == '0') {
```

Comment:
I would go for UNIX coreutils compatible tool UX here - so we might either give a human readable date or `@[unix_time]` as a parameter. For now we might want to support only unix timestamp, but adding `strptime`-based implementation should not be that hard in the future :).

Reply:
Implemented FORMAT string with suitable error message if used with `date -s` that this feature is not yet supported